File: backend/routers/customers.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, EmailStr
from typing import Optional
from database import execute_query

router = APIRouter()

class CustomerCreate(BaseModel):
    FirstName: str
    LastName: str
    Email: str
    NID: Optional[str] = None
    Gender: Optional[str] = None
    DoB: Optional[str] = None
    HouseNo: Optional[str] = None
    Area: Optional[str] = None
    City: Optional[str] = None
    District: Optional[str] = None
    PostalCode: Optional[str] = None
    PhoneNumbers: Optional[list[str]] = []

class CustomerUpdate(BaseModel):
    FirstName: Optional[str] = None
    LastName: Optional[str] = None
    Email: Optional[str] = None
    NID: Optional[str] = None
    Gender: Optional[str] = None
    DoB: Optional[str] = None
    HouseNo: Optional[str] = None
    Area: Optional[str] = None
    City: Optional[str] = None
    District: Optional[str] = None
    PostalCode: Optional[str] = None

@router.get("")
def get_customers(search: str = "", page: int = 1, limit: int = 10):
    offset = (page - 1) * limit
    if search:
        rows = execute_query(
            """SELECT c.*, GROUP_CONCAT(cp.PhoneNumber) as PhoneNumbers
               FROM customer c LEFT JOIN customerphone cp ON c.CustomerID=cp.CustomerID
               WHERE c.FirstName LIKE %s OR c.LastName LIKE %s OR c.Email LIKE %s OR c.City LIKE %s
               GROUP BY c.CustomerID ORDER BY c.CustomerID DESC LIMIT %s OFFSET %s""",
            (f"%{search}%", f"%{search}%", f"%{search}%", f"%{search}%", limit, offset)
        )
        count = execute_query(
            "SELECT COUNT(*) as total FROM customer WHERE FirstName LIKE %s OR LastName LIKE %s OR Email LIKE %s OR City LIKE %s",
            (f"%{search}%", f"%{search}%", f"%{search}%", f"%{search}%")
        )
    else:
        rows = execute_query(
            """SELECT c.*, GROUP_CONCAT(cp.PhoneNumber) as PhoneNumbers
               FROM customer c LEFT JOIN customerphone cp ON c.CustomerID=cp.CustomerID
               GROUP BY c.CustomerID ORDER BY c.CustomerID DESC LIMIT %s OFFSET %s""",
            (limit, offset)
        )
        count = execute_query("SELECT COUNT(*) as total FROM customer")
    return {"data": rows, "total": count[0]["total"], "page": page, "limit": limit}

@router.get("/{customer_id}")
def get_customer(customer_id: int):
    rows = execute_query(
        """SELECT c.*, GROUP_CONCAT(cp.PhoneNumber) as PhoneNumbers
           FROM customer c LEFT JOIN customerphone cp ON c.CustomerID=cp.CustomerID
           WHERE c.CustomerID=%s GROUP BY c.CustomerID""",
        (customer_id,)
    )
    if not rows:
        raise HTTPException(status_code=404, detail="Customer not found")
    return rows[0]

@router.post("", status_code=201)
def create_customer(customer: CustomerCreate):
    result = execute_query(
        """INSERT INTO customer (FirstName, LastName, Email, NID, Gender, DoB, RegDate, HouseNo, Area, City, District, PostalCode)
           VALUES (%s,%s,%s,%s,%s,%s,NOW(),%s,%s,%s,%s,%s)""",
        (customer.FirstName, customer.LastName, customer.Email, customer.NID,
         customer.Gender, customer.DoB, customer.HouseNo, customer.Area,
         customer.City, customer.District, customer.PostalCode),
        fetch=False
    )
    new_id = result["last_insert_id"]
    for phone in customer.PhoneNumbers:
        if phone:
            execute_query("INSERT INTO customerphone (CustomerID, PhoneNumber) VALUES (%s,%s)",
                          (new_id, phone), fetch=False)
    return {"message": "Customer created", "CustomerID": new_id}

@router.put("/{customer_id}")
def update_customer(customer_id: int, customer: CustomerUpdate):
    fields = {k: v for k, v in customer.dict().items() if v is not None}
    if not fields:
        raise HTTPException(status_code=400, detail="No fields to update")
    set_clause = ", ".join(f"{k}=%s" for k in fields)
    values = list(fields.values()) + [customer_id]
    execute_query(f"UPDATE customer SET {set_clause} WHERE CustomerID=%s", values, fetch=False)
    return {"message": "Customer updated"}

@router.delete("/{customer_id}")
def delete_customer(customer_id: int):
    execute_query("DELETE FROM customerphone WHERE CustomerID=%s", (customer_id,), fetch=False)
    execute_query("DELETE FROM customer WHERE CustomerID=%s", (customer_id,), fetch=False)
    return {"message": "Customer deleted"}

@router.get("/{customer_id}/bookings")
def get_customer_bookings(customer_id: int):
    return execute_query(
        """SELECT b.*, t.TourName, t.Price FROM booking b
           JOIN tourpackage t ON b.TourID=t.TourID
           WHERE b.CustomerID=%s ORDER BY b.BookingDate DESC""",
        (customer_id,)
    )

File: frontend/assets/js/pages/customers.js
let custPage = 1, custSearch = '';

async function renderCustomers() {
  const content = document.getElementById('page-content');
  content.innerHTML = `
    <div class="data-card">
      <div class="data-card-header">
        <h6><i class="bi bi-people me-2 text-primary"></i>Customers</h6>
        <div class="d-flex gap-2 flex-wrap">
          <div class="search-box">
            <i class="bi bi-search"></i>
            <input id="custSearch" type="text" class="form-control-custom" placeholder="Search customers..." oninput="debouncedCustSearch(this.value)" />
          </div>
          <button class="btn-primary-custom" onclick="openCustomerModal()"><i class="bi bi-plus-lg"></i> Add Customer</button>
        </div>
      </div>
      <div id="custTableWrap"><div class="loading-overlay"><div class="spinner"></div></div></div>
      <div class="p-3 d-flex justify-content-between align-items-center border-top" style="flex-wrap:wrap;gap:8px;">
        <small id="custInfo" class="text-muted"></small>
        <div id="custPagination"></div>
      </div>
    </div>

    <!-- Customer Modal -->
    <div class="modal fade" id="customerModal" tabindex="-1">
      <div class="modal-dialog modal-lg modal-dialog-centered">
        <div class="modal-content">
          <div class="modal-header">
            <h5 class="modal-title" id="custModalTitle">Add Customer</h5>
            <button type="button" class="btn-close" data-bs-dismiss="modal"></button>
          </div>
          <div class="modal-body">
            <div class="row g-3">
              <div class="col-md-6"><label class="form-label-custom">First Name *</label><input id="cf-fname" class="form-control-custom" placeholder="First name"></div>
              <div class="col-md-6"><label class="form-label-custom">Last Name *</label><input id="cf-lname" class="form-control-custom" placeholder="Last name"></div>
              <div class="col-md-6"><label class="form-label-custom">Email *</label><input id="cf-email" type="email" class="form-control-custom" placeholder="Email"></div>
              <div class="col-md-6"><label class="form-label-custom">NID</label><input id="cf-nid" class="form-control-custom" placeholder="National ID"></div>
              <div class="col-md-6">
                <label class="form-label-custom">Gender</label>
                <select id="cf-gender" class="form-control-custom"><option value="">Select</option><option>Male</option><option>Female</option><option>Other</option></select>
              </div>
              <div class="col-md-6"><label class="form-label-custom">Date of Birth</label><input id="cf-dob" type="date" class="form-control-custom"></div>
              <div class="col-12"><hr class="my-1"><p class="form-label-custom mb-2">Address</p></div>
              <div class="col-md-4"><label class="form-label-custom">House No</label><input id="cf-house" class="form-control-custom" placeholder="House No"></div>
              <div class="col-md-4"><label class="form-label-custom">Area</label><input id="cf-area" class="form-control-custom" placeholder="Area"></div>
              <div class="col-md-4"><label class="form-label-custom">City</label><input id="cf-city" class="form-control-custom" placeholder="City"></div>
              <div class="col-md-6"><label class="form-label-custom">District</label><input id="cf-district" class="form-control-custom" placeholder="District"></div>
              <div class="col-md-6"><label class="form-label-custom">Postal Code</label><input id="cf-postal" class="form-control-custom" placeholder="Postal Code"></div>
              <div class="col-12"><label class="form-label-custom">Phone Numbers (comma separated)</label><input id="cf-phones" class="form-control-custom" placeholder="01712345678, 01812345678"></div>
            </div>
          </div>
          <div class="modal-footer">
            <button class="btn btn-sm btn-secondary" data-bs-dismiss="modal">Cancel</button>
            <button class="btn-primary-custom" onclick="saveCustomer()"><i class="bi bi-check-lg"></i> Save</button>
          </div>
        </div>
      </div>
    </div>
  `;
  window.debouncedCustSearch = debounce(v => { custSearch = v; custPage = 1; loadCustomers(); });
  loadCustomers();
}

async function loadCustomers() {
  document.getElementById('custTableWrap').innerHTML = `<div class="loading-overlay"><div class="spinner"></div></div>`;
  try {
    const data = await apiGet('/customers', { search: custSearch, page: custPage, limit: 10 });
    if (!data.data.length) {
      document.getElementById('custTableWrap').innerHTML = `<div class="empty-state"><i class="bi bi-people"></i><p>No customers found</p></div>`;
      document.getElementById('custInfo').textContent = ''; return;
    }
    document.getElementById('custTableWrap').innerHTML = `
      <div class="table-wrapper">
        <table class="tms-table">
          <thead><tr><th>ID</th><th>Customer</th><th>Email</th><th>Phone</th><th>City</th><th>Gender</th><th>Reg Date</th><th>Actions</th></tr></thead>
          <tbody>
            ${data.data.map(c => `
              <tr>
                <td><span class="text-muted fw-500">#${c.CustomerID}</span></td>
                <td>
                  <div class="d-flex align-items-center gap-2">
                    <div class="avatar" style="background:${avatarColor(c.FirstName+c.LastName)};">${avatarInitials(c.FirstName+' '+c.LastName)}</div>
                    <div><div class="fw-600">${c.FirstName} ${c.LastName}</div><small class="text-muted">${c.NID || 'No NID'}</small></div>
                  </div>
                </td>
                <td>${c.Email}</td>
                <td>${c.PhoneNumbers ? c.PhoneNumbers.split(',')[0] : '-'}</td>
                <td>${c.City || '-'}</td>
                <td>${c.Gender || '-'}</td>
                <td>${fmtDate(c.RegDate)}</td>
                <td>
                  <div class="d-flex gap-1">
                    <button class="btn-icon btn-icon-view" title="View" onclick="viewCustomer(${c.CustomerID})"><i class="bi bi-eye"></i></button>
                    <button class="btn-icon btn-icon-edit" title="Edit" onclick="editCustomer(${c.CustomerID})"><i class="bi bi-pencil"></i></button>
                    <button class="btn-icon btn-icon-delete" title="Delete" onclick="deleteCustomer(${c.CustomerID})"><i class="bi bi-trash"></i></button>
                  </div>
                </td>
              </tr>`).join('')}
          </tbody>
        </table>
      </div>`;
    const start = (custPage - 1) * 10 + 1;
    const end = Math.min(custPage * 10, data.total);
    document.getElementById('custInfo').textContent = `Showing ${start}-${end} of ${data.total} customers`;
    buildPagination('custPagination', custPage, data.total, 10, p => { custPage = p; loadCustomers(); });
  } catch (e) {
    document.getElementById('custTableWrap').innerHTML = `<div class="empty-state"><i class="bi bi-exclamation-triangle"></i><p>${e.message}</p></div>`;
  }
}

let editingCustId = null;
function openCustomerModal(customer = null) {
  editingCustId = customer?.CustomerID || null;
  document.getElementById('custModalTitle').textContent = customer ? 'Edit Customer' : 'Add Customer';
  const fields = ['fname','lname','email','nid','gender','dob','house','area','city','district','postal','phones'];
  const mapping = { fname: customer?.FirstName, lname: customer?.LastName, email: customer?.Email, nid: customer?.NID, gender: customer?.Gender, dob: customer?.DoB?.split('T')[0], house: customer?.HouseNo, area: customer?.Area, city: customer?.City, district: customer?.District, postal: customer?.PostalCode, phones: customer?.PhoneNumbers };
  fields.forEach(f => { const el = document.getElementById('cf-'+f); if (el) el.value = mapping[f] || ''; });
  new bootstrap.Modal(document.getElementById('customerModal')).show();
}

async function editCustomer(id) {
  try {
    const c = await apiGet(`/customers/${id}`);
    openCustomerModal(c);
  } catch (e) { showToast(e.message, 'error'); }
}

async function viewCustomer(id) {
  try {
    const c = await apiGet(`/customers/${id}`);
    const bookings = await apiGet(`/customers/${id}/bookings`);
    // Show in a simple alert-style modal (reuse confirm modal for view)
    const info = `
      <b>${c.FirstName} ${c.LastName}</b> (ID: ${c.CustomerID})<br>
      Email: ${c.Email}<br>
      Phones: ${c.PhoneNumbers || 'N/A'}<br>
      Gender: ${c.Gender || 'N/A'} | DoB: ${fmtDate(c.DoB)}<br>
      Address: ${[c.HouseNo,c.Area,c.City,c.District,c.PostalCode].filter(Boolean).join(', ') || 'N/A'}<br>
      <br><b>Bookings (${bookings.length}):</b> ${bookings.map(b=>b.TourName).join(', ') || 'None'}
    `;
    document.getElementById('confirmMessage').innerHTML = info;
    document.getElementById('confirmBtn').style.display = 'none';
    const m = new bootstrap.Modal(document.getElementById('confirmModal'));
    document.getElementById('confirmModal').addEventListener('hidden.bs.modal', () => {
      document.getElementById('confirmBtn').style.display = '';
    }, { once: true });
    m.show();
  } catch (e) { showToast(e.message, 'error'); }
}

async function saveCustomer() {
  const phones = document.getElementById('cf-phones').value.split(',').map(p => p.trim()).filter(Boolean);
  const payload = {
    FirstName: document.getElementById('cf-fname').value.trim(),
    LastName: document.getElementById('cf-lname').value.trim(),
    Email: document.getElementById('cf-email').value.trim(),
    NID: document.getElementById('cf-nid').value.trim() || null,
    Gender: document.getElementById('cf-gender').value || null,
    DoB: document.getElementById('cf-dob').value || null,
    HouseNo: document.getElementById('cf-house').value.trim() || null,
    Area: document.getElementById('cf-area').value.trim() || null,
    City: document.getElementById('cf-city').value.trim() || null,
    District: document.getElementById('cf-district').value.trim() || null,
    PostalCode: document.getElementById('cf-postal').value.trim() || null,
    PhoneNumbers: phones,
  };
  if (!payload.FirstName || !payload.LastName || !payload.Email) { showToast('First name, last name, and email are required', 'warning'); return; }
  try {
    if (editingCustId) {
      await apiPut(`/customers/${editingCustId}`, payload);
      showToast('Customer updated successfully', 'success');
    } else {
      await apiPost('/customers', payload);
      showToast('Customer added successfully', 'success');
    }
    bootstrap.Modal.getInstance(document.getElementById('customerModal')).hide();
    loadCustomers();
  } catch (e) { showToast(e.message, 'error'); }
}

async function deleteCustomer(id) {
  const ok = await confirmDelete(`Delete customer #${id}? This will also remove their phone numbers.`);
  if (!ok) return;
  try {
    await apiDelete(`/customers/${id}`);
    showToast('Customer deleted', 'success');
    loadCustomers();
  } catch (e) { showToast(e.message, 'error'); }
}
